### DISCUSSION TITLE
How to print logs for unhandled events?

### DISCUSSION THREAD

**Discussion by Norlandz** [statelyai/xstate#4500] [Q&A]
When I use `actor.send(event)` && the `event` is not handled by the current state (-- because the current state is not listening for it).
I want to **print out an error log** for it, so that I know I did something wrong.

#### use `onEvent`
- this captures all events
- - to determine whether this event is appropriate (-- whether it will be handled by the (so-called) current state), 
    I tried to use **`currState.can(event)`**, 
    -- doesnt work 
  - -- because in Xstate the meaning of `currState (current state)` at `onEvent`
    is **not `the state before the event is sent (/ the state you are/were in, that will/would send the event)`**, 
    but is `the state after the event is sent`
    - (-- this is true (when I use things like `actions` too...))

#### use `strict mode`
- this captures some **internal events** like `['xstate.init', 'xstate.update', 'done.invokeXXXX']` (with `sync: true` with child state machine)
- this **throws** an error, not log
- deprecated

#### use wildcard event descriptor `'*'` 
- this captures some internal events like `['xstate.init', 'xstate.update', 'done.invokeXXXX']` (with `sync: true` with child state machine)
- if I just log the event without a target, seems like it goes to **infinite loop**
  if I just log the event with a target, the internal event will trigger this & **go to** that target state...
- I have to specify the wildcard in **every single state** 

### related question
https://stackoverflow.com/questions/58904591/what-happens-if-you-send-an-event-that-doesnt-exist-in-react-xstate

**Comment by Andarist** [comment]
I think that the wildcard event descriptor would be the best solution for you. 

> this captures some internal events like ['xstate.init', 'xstate.update', 'done.invokeXXXX'] (with sync: true with child state machine)

In the upcoming v5 all of the internal events are prefixed with `xstate.*` so you should be able to do smth like this:
```ts
'*': {
  actions: ({ event }) => {
    if (!event.type.startsWith('xstate.')) {
      console.log('Unhandled', event)
    }
  }
}
```

> if I just log the event without a target, seems like it goes to infinite loop

This sounds like some issue with your code. Maybe you combine this somehow with `always` transitions? I could take a look at the situation if you provide a runnable repro case of this problem.

> I have to specify the wildcard in every single state

You should be able to just put it on the root. Events "bubble" so if they stay unhandled by some inner states then they will reach the root and trigger the transition there.

---

In the future, we should be able to mark such unhandled events in the revamped devtools that we plan to build.

**Reply by Norlandz** [reply]
(Thanks for the reply, I will re-check my code to see if this works.)
